Problem of round-tripping Double values

rpetrusha opened the pull request:
# Problem of round-tripping Double values

## Summary

In come cases, the "R" format specifier fails to round-trip a Double value; our recommendation is to use "G17" instead. 

Fixes #2919

## Details

Some of the Double member documentation noted that "R" doesn't always successfully round-trip values and recommended "G17" instead. But the recommendation isn't consistent: In the standard numeric format strings, the Double structure, and the Single structure topics, "R" was recommended as the way to ensure a Double is round-tripped. This PR makes eliminates the inconsistency.

Internal builds are available as follows:
- [System.Double structure](https://review.docs.microsoft.com/en-us/dotnet/api/system.double?view=netframework-4.7&branch=pr-en-us-3101#Precision)
- [System.Single structure](https://review.docs.microsoft.com/en-us/dotnet/api/system.single?view=netframework-4.7&branch=pr-en-us-3101)
- [General format specifier, in Custom Numeric Format Strings](https://review.docs.microsoft.com/en-us/dotnet/standard/base-types/standard-numeric-format-strings?view=netframework-4.7&branch=pr-en-us-3101#the-general-g-format-specifier)
- [Round-trip format specifier, in Custom Numeric Format Strings](https://review.docs.microsoft.com/en-us/dotnet/standard/base-types/standard-numeric-format-strings?view=netframework-4.7&branch=pr-en-us-3101#the-round-trip-r-format-specifier)

## Suggested Reviewers
@mairaw @mazong1123 @tarekgh

Commit:
Problem of round-tripping Double values

tarekgh commented:
Is there a way I can look at nicely formatted doc?

mairaw commented:
Tes @tarekgh. We can post the link here once the build finishes.

mairaw commented:
@tarekgh here are the links to see the rendered topics:
- https://review.docs.microsoft.com/en-us/dotnet/standard/base-types/standard-numeric-format-strings?branch=pr-en-us-3101
- https://review.docs.microsoft.com/en-us/dotnet/api/system.double?branch=pr-en-us-3101&view=netframework-4.7#Remarks

rpetrusha commented:
@tarekgh, I've added links to the sections that I've modified in the Custom Numeric Format Strings topic in the PR description.

tarekgh commented on docs/standard/base-types/standard-numeric-format-strings.md:
```diff
@@ -52,15 +52,15 @@ Standard numeric format strings are used to format common numeric types. A stand
 |"N" or "n"|Number|Result: Integral and decimal digits, group separators, and a decimal separator with optional negative sign.<br /><br /> Supported by: All numeric types.<br /><br /> Precision specifier: Desired number of decimal places.<br /><br /> Default precision specifier: Defined by <xref:System.Globalization.NumberFormatInfo.NumberDecimalDigits%2A?displayProperty=fullName>.<br /><br /> Mo ...
 |"P" or "p"|Percent|Result: Number multiplied by 100 and displayed with a percent symbol.<br /><br /> Supported by: All numeric types.<br /><br /> Precision specifier: Desired number of decimal places.<br /><br /> Default precision specifier: Defined by  <xref:System.Globalization.NumberFormatInfo.PercentDecimalDigits%2A?displayProperty=fullName>.<br /><br /> More information: [The Percent ("P")  ...
-|"R" or "r"|Round-trip|Result: A string that can round-trip to an identical number.<br /><br /> Supported by: <xref:System.Single>, <xref:System.Double>, and <xref:System.Numerics.BigInteger>.<br /><br /> Precision specifier: Ignored.<br /><br /> More information: [The Round-trip ("R") Format Specifier](#RFormatString).|123456789.12345678 ("R") -> 123456789.12345678<br /><br /> -1234567890.123456 ...
+|"R" or "r"|Round-trip|Result: A string that can round-trip to an identical number.<br /><br /> Supported by: <xref:System.Single>, <xref:System.Double>, and <xref:System.Numerics.BigInteger>.<br /><br /> Note: Recommended for <xref:System.Single> and <xref:System.Numerics.BigInteger> types only; for <xref:System.Double> types, use "G17". </br> Precision specifier: Ignored.<br /><br /> More infor ...
```
><xref:System.Single>  [](start = 227, length = 21)

for Single we need to mention using "G9" as we mentioned G17 for double.

tannergooding commented:
For future reference, there is also a `rich-diff` view for markdown changes:
![image](https://user-images.githubusercontent.com/10487869/30293494-94aae38c-96ee-11e7-80eb-c03d9cacdee4.png)

tarekgh commented on docs/standard/base-types/standard-numeric-format-strings.md:
```diff
@@ -194,7 +194,12 @@ Standard numeric format strings are used to format common numeric types. A stand
  If scientific notation is used, the exponent in the result is prefixed with "E" if the format specifier is "G", or "e" if the format specifier is "g". The exponent contains a minimum of two digits. This differs from the format for scientific notation that is produced by the exponential format specifier, which includes a minimum of three digits in the exponent.  
-  
+ 
+Note that, when used with a <xref:System.Double> value, the "G17" format specifier ensures that the original <xref:System.Double> value successfully round-trips. This is because internally the runtime stores a <xref:System.Double> value with 17 digits of precision. We recommend its use instead of the ["R" format specifier](#RFormatString), since in some cases "R" fails to successfully round-trip  ...
```
>G17"  [](start = 61, length = 5)

You may mention G9 for Single too. you don't have to provide a code example for this one. 
I'll not mention G9 again in my comment as I expect you'll visit all places talking about G17 and add G9 part. thanks.

tannergooding commented on docs/standard/base-types/standard-numeric-format-strings.md:
```diff
@@ -52,15 +52,15 @@ Standard numeric format strings are used to format common numeric types. A stand
 |"N" or "n"|Number|Result: Integral and decimal digits, group separators, and a decimal separator with optional negative sign.<br /><br /> Supported by: All numeric types.<br /><br /> Precision specifier: Desired number of decimal places.<br /><br /> Default precision specifier: Defined by <xref:System.Globalization.NumberFormatInfo.NumberDecimalDigits%2A?displayProperty=fullName>.<br /><br /> Mo ...
 |"P" or "p"|Percent|Result: Number multiplied by 100 and displayed with a percent symbol.<br /><br /> Supported by: All numeric types.<br /><br /> Precision specifier: Desired number of decimal places.<br /><br /> Default precision specifier: Defined by  <xref:System.Globalization.NumberFormatInfo.PercentDecimalDigits%2A?displayProperty=fullName>.<br /><br /> More information: [The Percent ("P")  ...
-|"R" or "r"|Round-trip|Result: A string that can round-trip to an identical number.<br /><br /> Supported by: <xref:System.Single>, <xref:System.Double>, and <xref:System.Numerics.BigInteger>.<br /><br /> Precision specifier: Ignored.<br /><br /> More information: [The Round-trip ("R") Format Specifier](#RFormatString).|123456789.12345678 ("R") -> 123456789.12345678<br /><br /> -1234567890.123456 ...
+|"R" or "r"|Round-trip|Result: A string that can round-trip to an identical number.<br /><br /> Supported by: <xref:System.Single>, <xref:System.Double>, and <xref:System.Numerics.BigInteger>.<br /><br /> Note: Recommended for <xref:System.Single> and <xref:System.Numerics.BigInteger> types only; for <xref:System.Double> types, use "G17". </br> Precision specifier: Ignored.<br /><br /> More infor ...
```
> Recommended for <xref:System.Single>

I thought we had known cases where it could fail for `System.Single` as well?

tarekgh replied:
I already asked to fix that in my previous comments on this PR.

tannergooding replied:
Yeah, I started reviewing separately and saw you were tagging them after this comment 😄

tannergooding commented on docs/standard/base-types/standard-numeric-format-strings.md:
```diff
@@ -194,7 +194,12 @@ Standard numeric format strings are used to format common numeric types. A stand
  If scientific notation is used, the exponent in the result is prefixed with "E" if the format specifier is "G", or "e" if the format specifier is "g". The exponent contains a minimum of two digits. This differs from the format for scientific notation that is produced by the exponential format specifier, which includes a minimum of three digits in the exponent.  
-  
+ 
+Note that, when used with a <xref:System.Double> value, the "G17" format specifier ensures that the original <xref:System.Double> value successfully round-trips. This is because internally the runtime stores a <xref:System.Double> value with 17 digits of precision. We recommend its use instead of the ["R" format specifier](#RFormatString), since in some cases "R" fails to successfully round-trip  ...
```
> This is because internally the runtime stores a <xref:System.Double> value with 17 digits of precision.

Possibly:
> This is because <xref:System.Double> is an IEEE 754:2008 compliant double-precision floating-point number which gives up to 17 significant decimal digits of precision

Note to the above suggestion: The `IEEE 754:2008 double-precision floating-point` format is formally called `binary64`

rpetrusha replied:
@tannergooding, `Single` compiles with the standard as well (the binary32 format), no?

tannergooding replied:
Yes, and gives up to 9 significant decimal digits of precision.

tarekgh commented on docs/standard/base-types/standard-numeric-format-strings.md:
```diff
@@ -261,12 +266,15 @@ Standard numeric format strings are used to format common numeric types. A stand
-  
+ The round-trip ("R") format specifier attempts to ensure that a numeric value that is converted to a string is parsed back into the same numeric value. This format is supported only for the <xref:System.Single>, <xref:System.Double>, and <xref:System.Numerics.BigInteger> types.  
+
+For a <xref:System.Double> values, the "R" format specifier in some cases fails to successfully round-trip the original value. Instead, we recommend the that you use the ["G17"](#GFormatString) format specifier to successfully round-trip <xref:System.Double> values.
```
>the "R" format specifier in some cases fails to successfully round-trip the original value [](start = 35, length = 90)

This is a good comment but it may tell "R" is bad :-) can we a add a phrase telling like "R" try to optimize by formatting the number with 15 digit precisions which can fail the round tripping in some cases.

tannergooding replied:
@tarekgh, I'm not sure such a comment is useful. "R" technically is worse than "G17" in all cases. It is always less performant (due to converting back and checking if equal) and (in the cases of the known bugs) can be less accurate or completely inaccurate as well.

tarekgh replied:
Can we say, R try to format the number with 15 digits which in some cases cannot guarantee the round tripping?

tannergooding replied:
It's worth a separate discussion elsewhere, but I think if we are not going to be fixing "R" at this time, or in the future (due to back-compat) we should just deprecate it and tell people not to use it anymore because it is "bad" otherwise.

Users consuming "R" (outside of `BigInteger` where we don't think there are any bugs), are potentially introducing subtle bugs into their code and are taking a measurable performance penalty.

No matter how we word it, the issues will still exist unless we fix the underlying implementation.

rpetrusha replied:
@tarekgh, @tannergooding, I left the original wording and noted the performance issue ("in some cases fails to successfully round-trip the original value and also offers relatively poor performance.")

tarekgh replied:
I am fine with that if @tannergooding is fine too.

tannergooding replied:
That sounds reasonable to me 👍

tarekgh commented:
@rpetrusha I added a couple of comments. other than that it LGTM.

Commit:
Addressed review comments, revised Single documentation

rpetrusha commented:
@tarekgh, @tannergooding, I've addressed all of your comments and also made the necessary revisions to the documentation for the Single structure. Changes won't appear, though, until the build completes.

tarekgh commented:
:+1:

rpetrusha commented:
This doesn't look as if it ever built. Closing and reopening.

mairaw commented on docs/standard/base-types/standard-numeric-format-strings.md:
```diff
@@ -52,15 +52,15 @@ Standard numeric format strings are used to format common numeric types. A stand
  A standard numeric format string can be used to define the formatting of a numeric value in one of two ways:  
   
--   It can be passed to an overload of the `ToString` method that has a `format` parameter. The following example formats a numeric value as a currency string in the current (in this case, the en-US) culture.  
+-   It can be passed to an overload of the `ToString` method that has a `format` parameter. The following example formats a numeric value as a currency string in the current culture (in this case, the en-US culture).  
```
not related to this fix: line 75 
you an supply -> you can supply

mairaw commented on docs/standard/base-types/standard-numeric-format-strings.md:
```diff
@@ -194,7 +194,14 @@ Standard numeric format strings are used to format common numeric types. A stand
+[!code-csharp[Round-tripping a Double](../../../samples/snippets/standard/base-types/format-strings/csharp/g17.cs)]   
+[!code-vb[Round-tripping a Double](../../../samples/snippets/standard/base-types/format-strings/vb/g17.vb)]   
+
+When used with a <xref:System.Single> value, the "G9" format specifier ensures that the original <xref:System.Single> value successfully round-trips. This is because <xref:System.Single> is an IEEE 754-2008-compliant single-precision (`binary32`) floating point number that gives up to 9 signficant digits of precision. We recommend its use instead of the ["R" format specifier](#RFormatString), sin ...
```
typo: signficant

numeral: 9 -> nine

mairaw commented on docs/standard/base-types/standard-numeric-format-strings.md:
```diff
@@ -275,11 +283,11 @@ Standard numeric format strings are used to format common numeric types. A stand
 |<xref:System.Globalization.NumberFormatInfo.PositiveSign%2A>|Defines the string that indicates that an exponent is positive.|  
   
- The following example formats <xref:System.Double> values with the round-trip format specifier.  
+ The following example formats a <xref:System.BigInteger> value with the round-trip format specifier.  
```
broken link. should be System.Numerics.BigInteger

mairaw reviewed (approved):
LGTM. A few things to fix before merging.

Commit:
Incorporated reviewer comments

rpetrusha commented:
Closing and reopening to restart build.

Commit:
Merge branch 'master' into r-format

mairaw commented:
Looks good now @rpetrusha. Do you want to merge this?

rpetrusha commented:
@mairaw, I'll merge it now.

abelbraaksma commented on docs/standard/base-types/standard-numeric-format-strings.md:
```diff
@@ -261,12 +268,13 @@ Standard numeric format strings are used to format common numeric types. A stand
- Although you can include a precision specifier, it is ignored. Round trips are given precedence over precision when using this specifier.  
+ The round-trip ("R") format specifier attempts to ensure that a numeric value that is converted to a string is parsed back into the same numeric value. This format is supported only for the <xref:System.Single>, <xref:System.Double>, and <xref:System.Numerics.BigInteger> types.  
+
+For <xref:System.Double> and <xref:System.Single> values, the "R" format specifier in some cases fails to successfully round-trip the original value and also offers relatively poor performance. Instead, we recommend the that you use the ["G17"](#GFormatString) format specifier for <xref:System.Double> values and the ["G9"](#GFormatString) format specifier to successfully round-trip <xref:System.S ...
```
typo: `we recommend the that` -> `we recommend that`

abelbraaksma reviewed (commented):
Sorry, I know you are ready to merge, but just spotted one more typo.
